onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc

```diff
@@ -674,6 +676,119 @@ void RegisterLabelEncoderFeaturizerVer1() {
+  //     +-------+-------+-------------------------------------------------+
+  //     |Walmart| 9     | [[NAN, NAN], [ 10,  11], [NAN,   8], [ 9,  10]] |
+  //     +-------+-------+-------------------------------------------------+
+  //     |Walmart| 10    | [[NAN,   8], [ 11, NAN], [  8,   9], [10,  11]] |
+  //     +-------+-------+-------------------------------------------------+
+  //     |Walmart| 11    | [[  8,   9], [NAN, NAN], [  9,  10], [11, NAN]] |
+  //     +-------+-------+-------------------------------------------------+
+  //     Basically, if we have an offset of k for the row with row index t,
+  //     target_lag_k[t] = target[t - horizon + k + 1]
+  //)DOC";
+
+  MS_FEATURIZERS_OPERATOR_SCHEMA(LagLeadOperatorTransformer)
+      .SinceVersion(1)
+      .SetDomain(kMSFeaturizersDomain)
+      .Input(
+          0,
+          "State",
+          "State generated during training that is used for prediction",
+          "T0")
+      .Input(
+          1,
+          "Grains",
+          "Grains tensor of shape [R][K].",
+          "GrainT")
+      .Input(
+          2,
+          "Target",
+          "Target tensor of shape [R]",
+          "T")
+      .Output(
+          0,
+          "OutputGrains",
+          "Grains tensor of shape [R][K]",
+          "GrainT")
+      .Output(
+          1,
+          "Output",
+          "Output tensor of shape [R][P][Q]",
+          "OutputT")
+      .TypeConstraint(
+          "T0",
+          {"tensor(uint8)"},
+          "No information is available")
+      .TypeConstraint(
+          "GrainT",
+          {"tensor(string)"},
+          "No information is available")
+      .TypeConstraint(
+          "T",
+          {"tensor(int8)", "tensor(uint8)", "tensor(int16)",  "tensor(uint16)", "tensor(int32)", "tensor(uint32)", "tensor(int64)", "tensor(uint64)", "tensor(float)", "tensor(double)"},
+          "No information is available")
+      .TypeConstraint(
+          "OutputT",
+          {"tensor(double)"},
+          "No information is available")
+      .TypeAndShapeInferenceFunction(
+          [](ONNX_NAMESPACE::InferenceContext& ctx) {
+            propagateElemTypeFromInputToOutput(ctx, 1, 0);
+            propagateElemTypeFromDtypeToOutput(ctx, ONNX_NAMESPACE::TensorProto_DataType_DOUBLE, 1);
+            if (hasInputShape(ctx, 1) && hasInputShape(ctx, 2)) {
+              const auto& grains_shape = getInputShape(ctx, 1);
+              const auto& target_shape = getInputShape(ctx, 2);
+              if (grains_shape.dim_size() != 2) {
+                fail_shape_inference("Expecting Grains to have 2 dimensions");
+              }
+              if (target_shape.dim_size() != 1) {
+                fail_shape_inference("Expecting Target to have 1 dimensions");
+              }
+              propagateShapeFromInputToOutput(ctx, 1, 0);
+              ONNX_NAMESPACE::TensorShapeProto shape;
+              *shape.add_dim() = grains_shape.dim(0);
```

Comment:
This only requires grains shape to infer but if input 2 does not have shape this does not run.